lib/madwf_ml.cpp

```diff
@@ -0,0 +1,340 @@
+#include <madwf_ml.h>
+#include <madwf_transfer.h>
+
+namespace quda
+{
+
+  MadwfAcc::MadwfAcc(const SolverParam &solve_param, TimeProfile &profile) :
+    param(solve_param.madwf_param),
+    mu(param.madwf_diagonal_suppressor),
+    prec_precondition(solve_param.precision_precondition),
+    profile(profile)
+  {
+    if (getVerbosity() >= QUDA_VERBOSE) {
+      printfQuda("Launching MADWF accelerator ... \n");
+      printfQuda("madwf_mu (low modes suppressor)                   = %.4f\n", param.madwf_diagonal_suppressor);
+      printfQuda("madwf_ls (cheap Ls)                               = %d\n", param.madwf_ls);
+      printfQuda("madwf_null_miniter                                = %d\n", param.madwf_null_miniter);
+      printfQuda("madwf_null_tol                                    = %4.2e\n", param.madwf_null_tol);
+      printfQuda("madwf_train_maxiter (max # of iters for training) = %d\n", param.madwf_train_maxiter);
+    }
+  }
+
+  void MadwfAcc::fill_random(std::vector<transfer_float> &v)
+  {
+    static std::random_device rd;
+    // the good rng
+    static std::mt19937 rng(23ul * comm_rank());
+    // The gaussian distribution
+    static std::normal_distribution<double> n(0., 1.);
+
+    for (auto &x : v) { x = 1e-1 * n(rng); }
+  }
+
+  void MadwfAcc::apply(Solver &base, ColorSpinorField &out, const ColorSpinorField &in)
+  {
+    madwf_ml::transfer_5d_hh(*forward_tmp, in, device_param, false);
+    pushVerbosity(QUDA_SILENT);
+    base(*backward_tmp, *forward_tmp);
+    popVerbosity();
+    madwf_ml::transfer_5d_hh(out, *backward_tmp, device_param, true);
+
+    blas::axpy(mu, const_cast<ColorSpinorField &>(in), out);
+  }
+
+  double MadwfAcc::cost(const DiracMatrix &ref, Solver &base, ColorSpinorField &out, const ColorSpinorField &in)
+  {
+
+    ColorSpinorParam csParam(in);
+    ColorSpinorField tmp1(csParam);
+    ColorSpinorField tmp2(csParam);
+
+    apply(base, tmp1, in);
+    ref(tmp2, tmp1);
+
+    blas::copy(out, in);
+
+    // M * T^ * A * T * phi - phi
+    return blas::xmyNorm(tmp2, out);
+  }
+
+  void MadwfAcc::train(const DiracMatrix &ref, Solver &base, Solver &null, const ColorSpinorField &in,
+                       bool tune_suppressor)
+  {
+
+    constexpr int complex_matrix_size = static_cast<int>(transfer_t); // spin by spin
+
+    int Ls = in.X(4);
+    int Ls_base = param.madwf_ls;
+    int param_size = Ls * Ls_base * complex_matrix_size * 2;
+    std::vector<transfer_float> host_param(param_size);
+
+    if (param.madwf_param_load) {
+
+      profile.TPSTART(QUDA_PROFILE_IO);
+
+      char param_file_name[512];
```

Comment:
Maybe move this i/o handling to a separate function to improve readability?

Reply:
https://github.com/lattice/quda/pull/1061/commits/703f9d363082c78607d77ac0d576e66733264f35